File: src/test/java/com/everyonewaiter/domain/notification/AlimTalkWeblinkButtonTemplateTest.java
package com.everyonewaiter.domain.notification;

import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;

import org.junit.jupiter.api.Test;

class AlimTalkWeblinkButtonTemplateTest {

  @Test
  void getUri() {
    var template = AlimTalkWeblinkButtonTemplate.MENU_PREVIEW; // 필요 변수 1개

    assertThatCode(() -> template.createButton("value1"))
        .doesNotThrowAnyException();
  }

  @Test
  void getUriFail() {
    var template = AlimTalkWeblinkButtonTemplate.MENU_PREVIEW; // 필요 변수 1개

    assertThatThrownBy(() -> template.createButton("value1", "value2"))
        .isInstanceOf(IllegalArgumentException.class);
  }

}

File: src/test/java/com/everyonewaiter/domain/notification/AlimTalkWeblinkButtonTemplateTest.kt
package com.everyonewaiter.domain.notification

import com.everyonewaiter.domain.support.WordCounter
import org.assertj.core.api.Assertions.assertThatCode
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test

class AlimTalkWeblinkButtonTemplateTest {

  @Test
  fun `알림톡 버튼 생성`() {
    AlimTalkWeblinkButtonTemplate.entries.forEach {
      val parameterCount = WordCounter.count("%s", it.uri)
      if (parameterCount > 0) {
        assertThatCode {
          it.createButton(*Array(parameterCount) { "parameter" })
        }.doesNotThrowAnyException()
      } else {
        assertThatCode { it.createButton() }.doesNotThrowAnyException()
      }
    }
  }

  @Test
  fun `매개변수 개수가 옳바르지 않은 경우 알림톡 버튼 생성 실패`() {
    AlimTalkWeblinkButtonTemplate.entries.forEach {
      val parameterCount = WordCounter.count("%s", it.uri)
      assertThatThrownBy {
        it.createButton(*Array(parameterCount + 1) { "parameter" })
      }.isInstanceOf(IllegalArgumentException::class.java)
    }
  }
}
